test/test_torch.py

```diff
@@ -996,6 +996,10 @@ def test_unflatten(self):
             self.assertEqual(torch.tensor([1, 2, 3, 4]).unflatten(0, [2, 2]), torch.tensor([[1, 2], [3, 4]]))
             self.assertEqual(torch.tensor([1, 2, 3, 4]).unflatten(0, torch.Size([2, 2])), torch.tensor([[1, 2], [3, 4]]))
             self.assertEqual(torch.ones(2, 10).unflatten(1, (5, 2)), torch.ones(2, 5, 2))
+            self.assertEqual(torch.tensor([1, 2, 3, 4]).unflatten(0, (-1, 2)),
+                             torch.tensor([[1, 2], [3, 4]]))
+            self.assertEqual(torch.ones(2, 10).unflatten(1, (5, -1)),
+                             torch.ones(2, 5, 2))
```

Comment:
In both of these examples, we unflatten a single dimension into two dimensions. For completeness, what do you think about adding examples with more dimensions and an example with a single dimension? e.g.

```
torch.ones(2, 10).unflatten(1, (-1,))
torch.ones(2, 3*4*5*6).unflatten(1, (3, 4, -1, 6))
```

I'm also curious what happens if one tries to unflatten on a dimension of size 0 -- I imagine it raises an error but in either case it would be good to check the behavior

Reply:
Thanks so much for pointing this out, I have added the completeness test cases according to the kind suggestions.

As for a dimension of size 0, `unflatten` is able to infer the size for `-1` when there is no other size whose value is `0`.
```python
>>> torch.ones(2, 0, 2).unflatten(1, (3, -1, 4))
tensor([], size=(2, 3, 0, 4, 2))
```

Otherwise, it would raise an error of ambiguous values.
```python
>>> torch.ones(2, 0, 2).unflatten(1, (0, -1, 4))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/new_xiong/pytorch/torch/tensor.py", line 845, in unflatten
    return super(Tensor, self).unflatten(dim, sizes, names)
RuntimeError: cannot reshape tensor of 0 elements into shape [0, -1, 4] because the unspecified dimension size -1 can be any value and is ambiguous
```
The cases for testing a dimension of size 0 are added by the latest PR.

Reply:
Thank you for the update!